### PR TITLE
[ncp] add support for Commissioner APIs in spinel and NCP

### PR DESCRIPTION
**Pull request by abtink**
This commit adds support for Commissioner related APIs in spinel and
NCP including starting/stopping Commissioner, setting Provisioning
URL, adding/removing Joiners, and sending Announce Begin, Energy Scan
Query, PAN ID Query, MGMT_GET/MGMT_SET messages.

--------------
I have rebased this on top of branch from https://github.com/openthread/openthread/pull/2884 (so it have two commits)

### CODE REVIEW

**Comment by abtink** [comment]
In last push, rebased on master.

File: src/ncp/ncp_base_dispatcher.cpp [hunk 1]
```diff
@@ -948,6 +980,9 @@ NcpBase::PropertyHandler NcpBase::FindInsertPropertyHandler(spinel_prop_key_t aK
 
 #if OPENTHREAD_FTD
 #if OPENTHREAD_ENABLE_COMMISSIONER
+    case SPINEL_PROP_MESHCOP_COMMISSIONER_JOINERS:
```
Comment (librasungirl) [inline src/ncp/ncp_base_dispatcher.cpp]: Noticing that some of commissioner related code are wrapped with `OPENTHREAD_ENABLE_COMMISSIONER` and some are wrapped with `OPENTHREAD_FTD` as well. We might want to unify it.
MTD should also be commissioner-capable, right? @jwhui

File: src/ncp/spinel.h [hunk 2]
```diff
@@ -1589,10 +1602,170 @@ typedef enum {
...
+    /** Format `LCS6` - Write only
+     *
+     * Required capability: SPINEL_CAP_THREAD_COMMISSIONER
+     *
+     * Writing to this property sends an Announce Begin message with the specified parameters. Response is a
+     * `LAST_STATUS` update with status of operation.
+     *
+     *   `L` : Channel mask
+     *   `C` : Number of messages per channel
+     *   `S` : The time between energy measurements (milliseconds)
```
Comment (librasungirl) [inline src/ncp/spinel.h]: nit -> between announce transmissions

File: src/ncp/ncp_base_dispatcher.cpp [hunk 3]
```diff
@@ -1027,7 +1062,13 @@ NcpBase::PropertyHandler NcpBase::FindRemovePropertyHandler(spinel_prop_key_t aK
     case SPINEL_PROP_THREAD_ACTIVE_ROUTER_IDS:
         handler = &NcpBase::HandlePropertyRemove<SPINEL_PROP_THREAD_ACTIVE_ROUTER_IDS>;
         break;
+#if OPENTHREAD_ENABLE_COMMISSIONER
+    case SPINEL_PROP_MESHCOP_COMMISSIONER_JOINERS:
+        handler = &NcpBase::HandlePropertyRemove<SPINEL_PROP_MESHCOP_COMMISSIONER_JOINERS>;
+        break;
+        break;
```
Comment (librasungirl) [inline src/ncp/ncp_base_dispatcher.cpp]: nit: remove one `break` here

File: src/ncp/ncp_base_dispatcher.cpp (hunk 1 above)
Comment (abtink) [reply src/ncp/ncp_base_dispatcher.cpp]: Good point. Currently the commissioner APIs and implementation are tied to `FTD`.
- In the public APIs: [commissioner_api.cpp](https://github.com/openthread/openthread/blob/master/src/core/api/commissioner_api.cpp#L46)
- And in core implementation: [commissioner.cpp](https://github.com/openthread/openthread/blob/master/src/core/meshcop/commissioner.cpp#L57)

To be consistent, will use `#if OPENTHREAD_FTD && OPENTHREAD_ENABLE_COMMISSIONER` in NCP code for any commissioner related logic (where don't already have `FTD` check). Thanks

File: src/ncp/spinel.h [hunk 4]
```diff
@@ -1589,10 +1602,170 @@ typedef enum {
...
+     *
+     * Required capability: SPINEL_CAP_THREAD_COMMISSIONER
+     *
+     * This property provides asynchronous `CMD_PROP_VALUE_INSERTED` updates to report energy scan results for a
+     * previously sent Energy Scan Query message (please see `SPINEL_PROP_MESHCOP_COMMISSIONER_ENERGY_SCAN`).
+     *
+     * Format is:
+     *
+     *   `L` : Channel mask
+     *   `d` : Energy measurement data (not that `d` encoding includes the length)
```
Comment (librasungirl) [inline src/ncp/spinel.h]: nit: not->note?

File: src/ncp/ncp_base_dispatcher.cpp (hunk 1 above)
Comment (librasungirl) [reply src/ncp/ncp_base_dispatcher.cpp]: I am fine to keep the tie in this PR. We may raise a separate issue to discuss whether to remove such tie. Thanks.

**Pull request merged**